### PR TITLE
[spark] Support Spark 3.5

### LINKED ISSUE #920: [Feature] Support Spark 3.4
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

Apache Spark 3.4.0 is the fifth release of the 3.x line. Paimon could support the integration with Spark 3.4. The release note of Spark 3.4.0 is [Spark Release 3.4.0](https://spark.apache.org/releases/spark-release-3-4-0.html).

### Solution

Support Spark 3.4.

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### LINKED ISSUE #1995: [Feature] spark 3.5
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

spark 3.5: https://github.com/apache/spark/releases/tag/v3.5.0

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by SteNicholas**
### Purpose

Apache Spark 3.5.0 is the fifth release of the 3.x line. Paimon could support the integration with Spark 3.5. The release note of Spark 3.5.0 is [Spark Release 3.5.0](https://spark.apache.org/releases/spark-release-3-5-0.html). The following changes should be compatible between 3.5 and 3.4:

- `StructType` removes `toAttributes` method in [[SPARK-44353][CONNECT][SQL] Remove StructType.toAttributes](https://github.com/apache/spark/pull/41925).
- `RowEncoder` removes `apply` method in [[SPARK-44531][CONNECT][SQL] Move encoder inference to sql/api](https://github.com/apache/spark/pull/42134).

Fix #1995.

### Tests

paimon-spark-3.4 module adds `PaimonSinkTest`, `CreateAndDeleteTagProcedureTest` and `RollbackProcedureTest` to verify the compatibility of Spark 3.4.

### API and Format 

*(Does this change affect API or storage format)*

### Documentation

Document adds Spark 3.5 into the supported version of Spark.

### CODE REVIEW

**Comment by YannByron** [comment]
@SteNicholas @Zouxxyy is also working on supporting spark 3.5. I think this issue drives us to add some abstract design to address the api compatibility between spark versions.

**Comment by YannByron** [comment]
And, I recommend a single pr only focuses a single goal.

**Comment by SteNicholas** [comment]
@YannByron, the solution of the api compatibility has several ways including using the same api among versions. The 3.5 support uses the same api way to result in the compatibility. Therefore we could create another pull request to solve the api compatibility unifily and this pull request has no conflict with the api compatibility design.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala [hunk 1]
```diff
@@ -74,13 +75,13 @@ case class WriteIntoPaimonTable(
...
 
-    val dataEncoder = RowEncoder.apply(data.schema).resolveAndBind()
+    val dataEncoder = ExpressionEncoder(encoderFor(data.schema)).resolveAndBind()
     val originFromRow = dataEncoder.createDeserializer()
 
     // append _bucket_ column as placeholder
     val withBucketCol = data.withColumn(BUCKET_COL, lit(-1))
     val bucketColIdx = withBucketCol.schema.size - 1
-    val withBucketDataEncoder = RowEncoder.apply(withBucketCol.schema).resolveAndBind()
+    val withBucketDataEncoder = ExpressionEncoder(encoderFor(withBucketCol.schema)).resolveAndBind()
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala:84]: Can you check if this work for Spark 3.3 too?

Comment (SteNicholas) [reply paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala]: @JingsongLi, above implementation doesn't support Spark 3.3. I have used reflect way to solve compatibility problem and checked Spark 3.3 successful. PTAL.

**Comment by YannByron** [comment]
Can you make this pr focus only on supporting spark 3.5, no unit tests involved.

**Review by JingsongLi — APPROVED** [review]
Looks good to me!
At present, I see no harmful changes. Considering that this haven't been active for a while, I think it's worth merging.

**Pull request merged**